crates/red_knot/build.rs

```diff
@@ -0,0 +1,90 @@
+//! Build script to package our vendored typeshed files
+//! into a zip archive that can be included in the Ruff binary.
+//!
+//! This script should be automatically run at build time
+//! whenever the script itself changes, or whenever any files
+//! in `crates/red_knot/vendor/typeshed` change.
+
+use std::error::Error;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::Path;
+
+use walkdir::{DirEntry, WalkDir};
+use zip::result::ZipResult;
+use zip::write::{FileOptions, ZipWriter};
+use zip::CompressionMethod;
+
+const TYPESHED_SOURCE_DIR: &str = "vendor/typeshed";
+
+// NB: This is .gitignored; make sure to change the .gitignore entry
+// if you change this location
+const TYPESHED_ZIP_LOCATION: &str = "vendor/zipped_typeshed.zip";
+
+/// Recursively zip the contents of an entire directory.
+///
+/// This routine is adapted from a recipe at
+/// <https://github.com/zip-rs/zip-old/blob/5d0f198124946b7be4e5969719a7f29f363118cd/examples/write_dir.rs>
+fn zip_dir(
+    directory_iterator: &mut impl Iterator<Item = DirEntry>,
+    prefix: &str,
+    writer: File,
+) -> ZipResult<()> {
+    let mut zip = ZipWriter::new(writer);
+
+    let options = FileOptions::default()
+        .compression_method(CompressionMethod::Zstd)
+        .unix_permissions(0o644);
+
+    let mut buffer = Vec::new();
+    for entry in directory_iterator {
+        let path = entry.path();
+        let name = path
+            .strip_prefix(Path::new(prefix))
+            .unwrap()
+            .to_str()
+            .expect("Unexpected non-utf8 typeshed path!");
+
+        // Write file or directory explicitly
+        // Some unzip tools unzip files with directory paths correctly, some do not!
+        if path.is_file() {
+            println!("adding file {path:?} as {name:?} ...");
```

Comment:
Can we remove the println statements? Especially if they clutter the CLI output

Reply:
Because this is a build script, these aren't actually printed to stdout during the build process unless the build actually fails -- it's more like logging, really. But if the build fails, then the entire captured stdout from the build is printed in a summary, e.g.

![image](https://github.com/astral-sh/ruff/assets/66076021/45b61f59-998b-4460-a2be-4bf154d3ed5f)